File: fabrevealmenu/src/main/java/com/hlab/fabrevealmenu/view/FABMenuAdapter.java
package com.hlab.fabrevealmenu.view;

import android.content.res.ColorStateList;
import android.graphics.Typeface;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.hlab.fabrevealmenu.R;
import com.hlab.fabrevealmenu.helper.RevealDirection;
import com.hlab.fabrevealmenu.model.FABMenuItem;

import java.util.List;

public class FABMenuAdapter extends RecyclerView.Adapter<FABMenuAdapter.ViewHolder> {

    private FABRevealMenu parent;
    private List<FABMenuItem> mItems;
    private int rowLayoutResId = 0;
    private boolean showTitle = false;
    private boolean showIcon = true;
    private int titleTextColor;
    private int titleDisabledTextColor;
    private RevealDirection revealDirection;
    private boolean isCircularShape;
    private Typeface mMenuTitleTypeface;

    FABMenuAdapter(FABRevealMenu parent, List<FABMenuItem> mItems, int rowLayoutResId, boolean isCircularShape, int titleTextColor, int titleDisabledTextColor,
                   boolean showTitle, boolean showIcon, RevealDirection revealDirection) {
        this.parent = parent;
        this.mItems = mItems;
        this.rowLayoutResId = rowLayoutResId;
        this.isCircularShape = isCircularShape;
        this.showTitle = showTitle;
        this.showIcon = showIcon;
        this.titleTextColor = titleTextColor;
        this.titleDisabledTextColor = titleDisabledTextColor;
        this.revealDirection = revealDirection;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        return new ViewHolder(LayoutInflater.from(parent.getContext())
                .inflate(rowLayoutResId, parent, false));
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        holder.setData(mItems.get(position));
        holder.itemView.setEnabled(mItems.get(position).isEnabled());
        holder.tvTitle.setEnabled(mItems.get(position).isEnabled());
    }

    @Override
    public int getItemCount() {
        return mItems.size();
    }

    FABMenuItem getItemByIndex(int index) {
        if (index >= 0 && index < mItems.size()) {
            return mItems.get(index);
        }
        return null;
    }

    FABMenuItem getItemById(int id) {
        for (int i = 0; i < mItems.size(); i++) {
            if (mItems.get(i).getId() == id) {
                return mItems.get(i);
            }
        }
        return null;
    }

    void notifyItemChangedById(int id) {
        for (int i = 0; i < mItems.size(); i++) {
            if (mItems.get(i).getId() == id) {
                notifyItemChanged(i);
                return;
            }
        }
    }

    public boolean isShowTitle() {
        return showTitle;
    }

    public void setShowTitle(boolean showTitle) {
        this.showTitle = showTitle;
    }

    public boolean isShowIcon() {
        return showIcon;
    }

    public void setShowIcon(boolean showIcon) {
        this.showIcon = showIcon;
    }

    public int getTitleTextColor() {
        return titleTextColor;
    }

    public void setTitleTextColor(int titleTextColor) {
        this.titleTextColor = titleTextColor;
    }

    void setTitleDisabledTextColor(int titleDisabledTextColor) {
        this.titleDisabledTextColor = titleDisabledTextColor;
    }

    public RevealDirection getDirection() {
        return revealDirection;
    }

    void setDirection(RevealDirection revealDirection) {
        this.revealDirection = revealDirection;
    }

    void setMenuTitleTypeface(Typeface mMenuTitleTypeface) {
        this.mMenuTitleTypeface = mMenuTitleTypeface;
    }

    public class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {

        FABMenuItem item;
        TextView tvTitle;
        ImageView imgIcon;
        RelativeLayout viewParent;

        ViewHolder(View itemView) {
            super(itemView);
            viewParent = itemView.findViewById(R.id.view_parent);
            tvTitle = itemView.findViewById(R.id.txt_title_menu_item);
            tvTitle.setTextColor(new ColorStateList(new int[][]{new int[]{android.R.attr.state_enabled}, new int[]{-android.R.attr.state_enabled}}, new int[]{titleTextColor, titleDisabledTextColor}));
            tvTitle.setVisibility(showTitle ? View.VISIBLE : View.GONE);
            imgIcon = itemView.findViewById(R.id.img_menu_item);
            imgIcon.setVisibility(showIcon ? View.VISIBLE : View.GONE);

            if (mMenuTitleTypeface != null)
                tvTitle.setTypeface(mMenuTitleTypeface);

            viewParent.setBackgroundResource(isCircularShape ?
                    R.drawable.drawable_bg_selected_round : R.drawable.drawable_bg_selected);

            viewParent.setOnClickListener(this);
            viewParent.invalidate();
        }

        public void setData(FABMenuItem item) {
            this.item = item;
            viewParent.setTag(item.getId());
            tvTitle.setText(item.getTitle());
            imgIcon.setImageDrawable(item.getIconDrawable());

        }

        @Override
        public void onClick(View v) {
            if (item.isEnabled()) {
                parent.closeMenu();
                parent.menuSelectedListener.onMenuItemSelected(v, item.getId());
            }
        }
    }

}

File: fabrevealmenu/src/main/java/com/hlab/fabrevealmenu/view/FABMenuAdapter.kt
package com.hlab.fabrevealmenu.view

import android.content.res.ColorStateList
import android.graphics.Typeface
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.animation.AccelerateDecelerateInterpolator
import android.widget.ImageView
import android.widget.RelativeLayout
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.hlab.fabrevealmenu.R
import com.hlab.fabrevealmenu.enums.Direction
import com.hlab.fabrevealmenu.model.FABMenuItem

class FABMenuAdapter internal constructor(
        private val parent: FABRevealMenu,
        private val mItems: List<FABMenuItem>,
        private val rowLayoutResId: Int,
        private val isCircularShape: Boolean,
        var titleTextColor: Int,
        private var titleDisabledTextColor: Int,
        var showTitle: Boolean,
        var direction: Direction,
        var animateItems: Boolean
) : RecyclerView.Adapter<FABMenuAdapter.ViewHolder>() {

    companion object {
        private const val TRANSLATE_DISTANCE = 80
        private const val ANIM_DURATION = 350
        private const val ANIM_OFFSET = 50
        private const val ANIMATED_ITEMS_COUNT = 1
    }

    private val animInterpolator = AccelerateDecelerateInterpolator()
    private var isReturning = false
    private var lastAnimatedPosition = -1
    private var maxDuration: Int = 0
    private var mMenuTitleTypeface: Typeface? = null

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        return ViewHolder(LayoutInflater.from(parent.context)
                .inflate(rowLayoutResId, parent, false))
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.setData(mItems[position])
        holder.itemView.isEnabled = mItems[position].isEnabled
        holder.tvTitle.isEnabled = mItems[position].isEnabled
        // Here you apply the animation when the view is bound
        runEnterAnimation(holder.itemView, position)
    }

    override fun getItemCount(): Int {
        return mItems.size
    }

    internal fun getItemByIndex(index: Int): FABMenuItem? {
        return if (index >= 0 && index < mItems.size) {
            mItems[index]
        } else null
    }

    internal fun getItemById(id: Int): FABMenuItem? {
        for (i in mItems.indices) {
            if (mItems[i].id == id) {
                return mItems[i]
            }
        }
        return null
    }

    internal fun notifyItemChangedById(id: Int) {
        for (i in mItems.indices) {
            if (mItems[i].id == id) {
                notifyItemChanged(i)
                return
            }
        }
    }

    private fun runEnterAnimation(view: View, position: Int) {
        if (!animateItems || position < ANIMATED_ITEMS_COUNT - 1) {
            return
        }

        if (position > lastAnimatedPosition) {
            lastAnimatedPosition = position
            if (isReturning)
                startExitAnimation(view, position)
            else
                startEnterAnimation(view, position)
        }
    }

    internal fun resetAdapter(returning: Boolean) {
        lastAnimatedPosition = -1
        isReturning = returning
        notifyDataSetChanged()
        maxDuration = itemCount * ANIM_OFFSET
    }

    private fun startEnterAnimation(view: View, position: Int) {
        val translateTo = if (direction == Direction.DOWN) -TRANSLATE_DISTANCE else TRANSLATE_DISTANCE
        view.translationY = translateTo.toFloat()
        view.alpha = 0f
        view.animate()
                .translationY(0f)
                .alpha(1f)
                .setInterpolator(animInterpolator)
                .setDuration(ANIM_DURATION.toLong())
                .setStartDelay((ANIM_OFFSET + position * ANIM_OFFSET).toLong())
                .start()
    }

    private fun startExitAnimation(view: View, position: Int) {
        val translateTo = if (direction == Direction.DOWN) -TRANSLATE_DISTANCE else TRANSLATE_DISTANCE
        view.translationY = 0f
        view.alpha = 1f
        view.animate()
                .translationY(translateTo.toFloat())
                .alpha(0f)
                .setInterpolator(animInterpolator)
                .setDuration(ANIM_DURATION.toLong())
                .setStartDelay((maxDuration - position * ANIM_OFFSET).toLong())
                .start()
    }

    internal fun setTitleDisabledTextColor(titleDisabledTextColor: Int) {
        this.titleDisabledTextColor = titleDisabledTextColor
    }

    internal fun setMenuTitleTypeface(mMenuTitleTypeface: Typeface) {
        this.mMenuTitleTypeface = mMenuTitleTypeface
    }

    inner class ViewHolder internal constructor(itemView: View) : RecyclerView.ViewHolder(itemView), View.OnClickListener {

        private var item: FABMenuItem? = null
        internal var tvTitle: TextView = itemView.findViewById(R.id.txt_title_menu_item)
        private var imgIcon: ImageView = itemView.findViewById(R.id.img_menu_item)
        private var viewParent: RelativeLayout = itemView.findViewById(R.id.view_parent)

        init {
            tvTitle.setTextColor(ColorStateList(arrayOf(intArrayOf(android.R.attr.state_enabled), intArrayOf(-android.R.attr.state_enabled)), intArrayOf(titleTextColor, titleDisabledTextColor)))
            tvTitle.visibility = if (showTitle) View.VISIBLE else View.GONE


            if (mMenuTitleTypeface != null)
                tvTitle.typeface = mMenuTitleTypeface

            viewParent.setBackgroundResource(if (isCircularShape)
                R.drawable.drawable_bg_selected_round
            else
                R.drawable.drawable_bg_selected)

            viewParent.setOnClickListener(this)
            viewParent.invalidate()
        }

        fun setData(item: FABMenuItem) {
            this.item = item
            viewParent.tag = item.id
            tvTitle.text = item.title
            imgIcon.setImageDrawable(item.iconDrawable)

        }

        override fun onClick(v: View) {
            if (item?.isEnabled == true) {
                parent.closeMenu()
                parent.menuSelectedListener?.onMenuItemSelected(v, item!!.id)
            }
        }
    }

}
